client/client_session.cpp

```diff
@@ -1,70 +1,151 @@
+                // TODO maybe just remove the catch
+                std::cerr << "fatal error: " << ex.what() << "\n";
+                break;
+                // throw ex;
+            }
+        }
+    });
 }
 
-nibaclient::client_session::~client_session() { close(); }
+void nibaclient::client_session::block_until_ready() { ready_future_.get(); }
 
-void nibaclient::client_session::close() {
-    if (ws_.is_open()) {
-        ws_.close(websocket::close_code::normal);
-    }
-    timer_.cancel();
+void nibaclient::client_session::stop() {
+    // Exceptions may happen in the destructor, but if it throws, then let it crash...?
+    boost::asio::spawn(ioc_, [this, self = shared_from_this()](boost::asio::yield_context yield) {
+        if (ws_.is_open()) {
+            // ioc should finish the close before shutting down
+            beast::error_code ec;
+            ws_.async_close(websocket::close_code::normal, yield[ec]);
+            if (ec && ec != boost::asio::error::operation_aborted) {
+                std::cerr << "Close error: " << ec.message() << std::endl;
+            }
+        }
+    });
 }
 
-void nibaclient::client_session::handle_cmd(const std::string &input) {
+std::future<nibashared::message::type>
+nibaclient::client_session::handle_cmd(const std::string &input) {
     try {
-        if (input == "exit") {
-            std::cout << "goodbye" << std::endl;
-            // this might not be the last command... but works for tests
-            close();
-            ioc_.stop();
-            return;
-        }
         handled_ = true;
         std::vector<std::string> results;
         boost::split(results, input, boost::is_any_of("\t "));
         if (results.empty())
             throw std::runtime_error("empty input");
-        // dynamic sized
+        // Dynamically sized arguments
         if (results[0] == "reordermagic") {
             std::vector<int> selected;
-            std::for_each(results.begin() + 1, results.end(), [&selected](auto &magic_id) {
-                selected.push_back(std::stoi(magic_id));
-            });
+            std::for_each(results.begin() + 1, results.end(),
+                          [&selected](auto &magic_id) { selected.push_back(std::stoi(magic_id)); });
             return create_and_go<nibashared::message_reordermagic>(std::move(selected));
+        } else if (results[0] == "send") {
+            if (results.size() < 3) {
+                throw std::runtime_error("need more arguments");
+            }
+            std::stringstream ss;
+            for (auto iter = std::next(results.begin(), 2); iter != results.end(); ++iter) {
+                ss << *iter;
+                if (iter != std::prev(results.end())) {
+                    ss << " ";
+                }
+            }
+            return create_and_go<nibashared::message_send>(std::move(results[1]), ss.str());
         }
-        // password input handled by cmd_processor
         if (results.size() == 3) {
```

Comment:
Is there a way to restructure the
```
if
  if
else if
  if
```
logic? If we end up having 100 commands, it will be hard to maintain.

Reply:
not sure if there is a good way, some crazy template meta programming might do... we could find a library to use if we go down this route, but I think commandline handling is mostly for testing, eventually we'll find a better way I think